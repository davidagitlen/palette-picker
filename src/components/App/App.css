h1 {
  background-color: #424A60;
  display: inline;
  color: #C8CAC8;
  font-family: 'Source Code Pro', monospace;
  font-size: 66px;
  line-height: 88px;
  margin: 0px 0px 25px 0px;

}

.App {
  display: flex;
  flex-direction: column;
  text-align: center;
  background-color: ghostwhite;
}

.top-wrapper {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  margin: 10px 70px;
}

.controls-background {
  background-color: rgba(240,240,240, 0.7);
  box-shadow: 1px 1px 4px #D3D3D3;
}