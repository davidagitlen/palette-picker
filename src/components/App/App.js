import React, { Component } from 'react';
import './App.css';
import Display from '../Display/Display';
import Controls from '../Controls/Controls';
import Preview from "../Preview/Preview";
import Projects from '../Projects/Projects';
import { 
  getProjects,
  getPalettes,
  postPalette,
  postProject,
  patchProject,
  patchPalette,
  deleteProject,
  deletePalette,
  searchByHex
  } from '../../util/apiCalls';
import {
  generateHex,
  lockColors,
  mapLockedColors,
  reconstructColors
  } from '../../util/dataHandlers';
import headerRt from '../../images/header-rt.png';
import headerLt from '../../images/header-lt.png';

class App extends Component {
  constructor() {
    super();
    this.state = {
      currentPalette: {
        palette: "",
        project_name: "",
        colors: [
          { hex_1: generateHex(), locked: false },
          { hex_2: generateHex(), locked: false },
          { hex_3: generateHex(), locked: false },
          { hex_4: generateHex(), locked: false },
          { hex_5: generateHex(), locked: false }
        ]
      },
      projects: [],
      palettes: [],
      error: "",
      foundPalettes: []
    };
  }

  componentDidMount = async () => {
    const projects = await getProjects();
    this.setState({ projects });
    const palettes = await getPalettes();
    this.setState({ palettes });
  };

  getRandomHexes = () => {
    const { colors, project_name, palette } = this.state.currentPalette;
    const newColors = mapLockedColors(colors);
    this.setState({ currentPalette: { palette, project_name, colors: newColors }});
  };

  saveCurrentPalette = async (paletteName, projectName) => {
    const palette = paletteName;
    const project_name = projectName;
    const { projects } = this.state;
    const { colors } = this.state.currentPalette;
    await this.setState({currentPalette: {palette, project_name, colors}});
    if (projects.some(project => project.project === projectName)) {
      this.handlePalettePost();
    } else {
      this.handlePaletteAndProjectPost(projectName);
    }
  };

  handlePalettePost = async () => {
    const { currentPalette } = this.state;
    try {
      await postPalette(currentPalette);
      const palettes = await getPalettes();
      this.setState({ palettes });
    } catch (error) {
      this.setState({ error });
    }
  }

  handlePaletteAndProjectPost = async (projectName) => {
    const { currentPalette } = this.state;
    try {
      await postProject(projectName);
      await postPalette(currentPalette);
      const projects = await getProjects();
      const palettes = await getPalettes();
      this.setState({ projects, palettes });
    } catch (error) {
      this.setState({ error });
    }
  }

  toggleLock = position => {
    const { colors, project_name, palette } = this.state.currentPalette;
    const newColors = lockColors(colors, position);
    this.setState({ currentPalette: { palette, project_name, colors: newColors } });
  };

  editProject = async (projectName, id) => {
    try {
      await patchProject(projectName, id);
      const projects = await getProjects();
      this.setState({ projects });
    } catch (error) {
      this.setState({ error });
    }
  };

  trashProject = async id => {
    try {
      await deleteProject(id);
      const projects = await getProjects();
      this.setState({ projects });
    } catch (error) {
      this.setState({ error });
    }
  };

  editPalette = async (palette, id) => {
    try {
      await patchPalette(palette, id);
      const palettes = await getPalettes();
      this.setState({ palettes });
    } catch (error) {
      this.setState({ error });
    }
  };

  trashPalette = async id => {
    try {
      await deletePalette(id);
      const palettes = await getPalettes();
      this.setState({ palettes });
    } catch (error) {
      this.setState({ error });
    }
  };

  selectPalette = selectedPalette => {
    const { palette, project_name } = selectedPalette;
    const colors = reconstructColors(selectedPalette);
    this.setState({currentPalette: { palette , project_name, colors}});
    window.scrollTo({top: 0, left: 0, behavior: 'smooth'})
  }

  findPalettes = async (e, hex) => {
    e.preventDefault();
    try {
      const foundPalettes = await searchByHex(hex);
      this.setState({foundPalettes});
    } catch (error) {
      this.setState({error})
    }
  }

  clearSearch = () => {
    this.setState({foundPalettes: []});
  }

  render() {
    const { currentPalette, foundPalettes } = this.state;
    let foundProjects;
    if (foundPalettes.length) {
      foundProjects = this.state.projects.filter(project => 
        foundPalettes.some(palette => palette.project_id === project.id)
      )
    }
    const palettes = foundPalettes.length ? foundPalettes : this.state.palettes;
    const projectsToShow = foundPalettes.length ? foundProjects : this.state.projects;
    return (
      <main className="App">
        <header>
          <img src={headerLt} alt="logo-left" className="header-img" />
          <h1>Color Schema</h1>
          <img src={headerRt} alt="logo-right" className="header-img" />
        </header>
        <div className="top-wrapper">
          <div className="controls-background">
            <Controls
              projects={this.state.projects}
              saveCurrentPalette={this.saveCurrentPalette}
              getRandomHexes={this.getRandomHexes}
              findPalettes={this.findPalettes}
              clearSearch={this.clearSearch}
            />
          </div>
          <Display
            currentPalette={currentPalette}
            toggleLock={this.toggleLock}
          />
          <Preview
            projectName={currentPalette.project_name}
            paletteName={currentPalette.palette}
            colors={currentPalette}
          />
        </div>
        <Projects
          projects={projectsToShow}
          palettes={palettes}
          editProject={this.editProject}
          editPalette={this.editPalette}
          trashProject={this.trashProject}
          trashPalette={this.trashPalette}
          selectPalette={this.selectPalette}
        />
      </main>
    );
  }
}

export default App;