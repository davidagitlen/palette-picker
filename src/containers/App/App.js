import React, { Component } from 'react';
import './App.css';
import Display from '../Display/Display';
import Controls from '../Controls/Controls';
import Projects from '../Projects/Projects';
import { getProjects, getPalettes } from '../../util/apiCalls'; 

class App extends Component{
  constructor(){
    super();
    this.state = {
      currentPalette : {
        palette: '',
        project_name: '',
        colors: [
          { hex_1: this.generateHex(), locked: false },
          { hex_2: this.generateHex(), locked: false },
          { hex_3: this.generateHex(), locked: false },
          { hex_4: this.generateHex(), locked: false },
          { hex_5: this.generateHex(), locked: false },
        ]
      },
      projects: [],
      palettes: []
    }
  }

  componentDidMount = async () => {
    const projects = await getProjects();
    this.setState({projects}, () => {console.log(this.state)})
    const palettes = await getPalettes();
    this.setState({palettes}, () => {console.log(this.state.currentPalette.colors)})
  }
  
  generateHex = (string = '#') => {
    const hexadecimals = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'];
    const index = Math.floor(Math.random() * 16);
    if (string.length === 7) {
      return string;
    } else {
      return this.generateHex(string + hexadecimals[index])
    }
  }

  getRandomHexes = () => {
    const { colors, project_name, palette } = this.state.currentPalette; 
    const newColors = colors.map((color, i) => {
      if (!color.locked) {
        color[`hex_${i+1}`] = this.generateHex();
        return color;
      } else {
        return color;
      }
    });
    this.setState({currentPalette : { palette, project_name, colors: newColors } }, () => {console.log(this.state)})
  }

  saveCurrentPalette = (paletteName, projectName) => {
    const { colors }  = this.state.currentPalette;
    this.setState({ currentPalette: { palette: paletteName, project_name: projectName, colors: colors } }, () => { console.log('after save', this.state) });
  }

  toggleLock = (position) => {
    const { colors, project_name, palette } = this.state.currentPalette;
    const newColors = colors.map((color, i) => {
      if (i === position) {
        color.locked = !color.locked;
        return color;
      } else {
        return color;
      }
    });
    this.setState({ currentPalette: { palette, project_name, colors: newColors } }, () => { console.log(this.state) })
  }

  render(){
    const { currentPalette, projects, palettes } = this.state;
    return (
      <main className="App">
        <h1>App</h1>
        <Display 
          currentPalette={currentPalette}
          toggleLock={this.toggleLock}/>
        <Controls 
          projects={projects} 
          saveCurrentPalette={this.saveCurrentPalette}
          getRandomHexes={this.getRandomHexes}
          />
        <Projects />
      </main>
    );
  }
}

export default App;
