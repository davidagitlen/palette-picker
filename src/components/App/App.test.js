import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  getProjects,
  getPalettes,
  postPalette,
  postProject
} from '../../util/apiCalls';

jest.mock('../../util/apiCalls');

configure({ adapter: new Adapter() });

describe('App', () => {
  let wrapper, mockState;
  
  beforeEach(() => {
    wrapper = shallow(<App/>);
    mockState = {
      currentPalette: {
        palette: "Mock_Palette",
        project_name: "Mock_Project",
        colors: [
          { hex_1: '#ABABAB', locked: false },
          { hex_2: '#CDCDCD', locked: false },
          { hex_3: '#EFEFEF', locked: false },
          { hex_4: '#121212', locked: false },
          { hex_5: '#343434', locked: false }
        ]
      },
      projects: [
        {id: 1, project: 'Project_One'}, 
        {id: 2, project: 'Project_Two'}
      ],
      palettes: [
        {
          id: 1, 
          project_name: 'Project_One', 
          hex_1: '#565656',
          hex_2: '#787878',
          hex_3: '#909090',
          hex_4: '#A1A1A1',
          hex_5: '#B2B2B2',
          project_id: 1
        },
        {
          id: 2,
          project_name: 'Project_Two',
          hex_1: '#565656',
          hex_2: '#787878',
          hex_3: '#909090',
          hex_4: '#A1A1A1',
          hex_5: '#B2B2B2',
          project_id: 2
        }
      ],
      error: "",
      foundPalettes: []
    };
  });


  it('should match the snapshot', () => {
    wrapper.instance().setState(mockState);
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('should fire getProjects', () => {
      expect(getProjects).toHaveBeenCalled();
    });

    it('should fire getPalettes', () => {
      expect(getPalettes).toHaveBeenCalled();
    });
  });

  describe('getRandomHexes', () => {
    it('should return an array of new random hex colors and set it to state', () => {
      const { colors } = wrapper.state().currentPalette;
      wrapper.instance().getRandomHexes();
      const newColors = wrapper.state().currentPalette
      expect(newColors).not.toEqual(colors);
    });
  });

  describe('saveCurrentPalette', () => {
    it('should update state with the palette and project names passed into it', async () => {
      wrapper.instance().setState(mockState);
      const { palette, project_name } = wrapper.state('currentPalette');
      expect(palette).toEqual('Mock_Palette');
      expect(project_name).toEqual('Mock_Project');
      await wrapper.instance().saveCurrentPalette('New_Palette', 'New_Name');
      const { palette: newPalette, project_name: new_Project_Name } = wrapper.state('currentPalette');
      expect(newPalette).toEqual('New_Palette');
      expect(new_Project_Name).toEqual('New_Name');
    }); 

    it('should invoke handlePalettePost if passed a value of a projectName that already exists in state', async () => {
      wrapper.instance().setState(mockState);
      wrapper.instance().handlePalettePost = jest.fn();
      await wrapper.instance().saveCurrentPalette('Test_Palette', 'Project_One');
      expect(wrapper.instance().handlePalettePost).toHaveBeenCalled();
    });

    it('should invoke both handlePalettePost and handlePaletteAndProjectPost if passed a value of a projectName that does not already exist in state', async () => {
      wrapper.instance().setState(mockState);
      wrapper.instance().handlePaletteAndProjectPost = jest.fn();
      await wrapper.instance().saveCurrentPalette('Test_Palette', 'New_Project_Name');
      expect(wrapper.instance().handlePaletteAndProjectPost).toHaveBeenCalled();
    });
  });

  describe('handlePalettePost', () => {
    it('should invoke postPalette with the currentPalette passed in', async () => {
      const { currentPalette } = wrapper.state(); 
      await wrapper.instance().handlePalettePost();
      expect(postPalette).toHaveBeenCalledWith(currentPalette);
    });
  });

  describe('handlePaletteAndProjectPost', () => {
    it('should invoke postProject with the current projectName passed in', async () => {
      const { project_name } = wrapper.state();
      await wrapper.instance().handlePaletteAndProjectPost();
      expect(postProject).toHaveBeenCalledWith(project_name);
    });

    it('should invoke postPalette with the currentPalette passed in', async () => {
      const { currentPalette } = wrapper.state();
      await wrapper.instance().handlePalettePost();
      expect(postPalette).toHaveBeenCalledWith(currentPalette);
    });
  });

});
