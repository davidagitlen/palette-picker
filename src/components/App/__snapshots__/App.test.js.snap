// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<main
  className="App"
>
  <h1>
    Color Schema
  </h1>
  <div
    className="top-wrapper"
  >
    <div
      className="controls-background"
    >
      <Controls
        clearSearch={[Function]}
        findPalettes={[Function]}
        getRandomHexes={[Function]}
        projects={
          Array [
            Object {
              "id": 1,
              "project": "Project_One",
            },
            Object {
              "id": 2,
              "project": "Project_Two",
            },
          ]
        }
        saveCurrentPalette={[Function]}
      />
    </div>
    <Display
      currentPalette={
        Object {
          "colors": Array [
            Object {
              "hex_1": "#ABABAB",
              "locked": false,
            },
            Object {
              "hex_2": "#CDCDCD",
              "locked": false,
            },
            Object {
              "hex_3": "#EFEFEF",
              "locked": false,
            },
            Object {
              "hex_4": "#121212",
              "locked": false,
            },
            Object {
              "hex_5": "#343434",
              "locked": false,
            },
          ],
          "palette": "Mock_Palette",
          "project_name": "Mock_Project",
        }
      }
      toggleLock={[Function]}
    />
    <Preview
      colors={
        Object {
          "colors": Array [
            Object {
              "hex_1": "#ABABAB",
              "locked": false,
            },
            Object {
              "hex_2": "#CDCDCD",
              "locked": false,
            },
            Object {
              "hex_3": "#EFEFEF",
              "locked": false,
            },
            Object {
              "hex_4": "#121212",
              "locked": false,
            },
            Object {
              "hex_5": "#343434",
              "locked": false,
            },
          ],
          "palette": "Mock_Palette",
          "project_name": "Mock_Project",
        }
      }
      paletteName="Mock_Palette"
      projectName="Mock_Project"
    />
  </div>
  <Projects
    editPalette={[Function]}
    editProject={[Function]}
    palettes={
      Array [
        Object {
          "hex_1": "#565656",
          "hex_2": "#787878",
          "hex_3": "#909090",
          "hex_4": "#A1A1A1",
          "hex_5": "#B2B2B2",
          "id": 1,
          "project_id": 1,
          "project_name": "Project_One",
        },
        Object {
          "hex_1": "#565656",
          "hex_2": "#787878",
          "hex_3": "#909090",
          "hex_4": "#A1A1A1",
          "hex_5": "#B2B2B2",
          "id": 2,
          "project_id": 2,
          "project_name": "Project_Two",
        },
      ]
    }
    projects={
      Array [
        Object {
          "id": 1,
          "project": "Project_One",
        },
        Object {
          "id": 2,
          "project": "Project_Two",
        },
      ]
    }
    selectPalette={[Function]}
    trashPalette={[Function]}
    trashProject={[Function]}
  />
</main>
`;
